File: src/crtp/crtp/packers/parameters_packer.py
import struct

from .toc_packer import TocPacker

class ParametersPacker(TocPacker):
    PORT_PARAMETER = 0x02
    
    TOC_CHANNEL = 0
    READ_CHANNEL = 1
    WRITE_CHANNEL = 2
    MISC_CHANNEL = 3


    def __init__(self, CrtpPacker):
        super().__init__(CrtpPacker, self.PORT_PARAMETER)
    
    def set_parameter(self, id, pytype, value):
        data = struct.pack('<H', id)
        data += struct.pack(pytype, value)
        return self._prepare_packet(self.WRITE_CHANNEL, data)
        

File: src/crtp_cpp/src/crtp_cpp/packer/parameters_packer.cpp

#include "crtp_cpp/packer/parameters_packer.hpp"
#include <cstring>

#define PORT_PARAMETER 0x02

#define TOC_CHANNEL 0
#define READ_CHANNEL 1
#define WRITE_CHANNEL 2
#define MISC_CHANNEL 3

ParametersPacker::ParametersPacker()
    : TocPacker(PORT_PARAMETER) {}

CrtpPacket ParametersPacker::prepare_packet(const std::vector<uint8_t>& data, uint8_t channel) {
    return CrtpPacker::prepare_packet(channel, data);
}

CrtpRequest ParametersPacker::set_parameter(uint16_t id, ParamValue value, ParamType type) {
    CrtpRequest request;
    std::vector<uint8_t> data(2); // ID (uint16_t)
    // TODO: To Be implemented
    request.packet = prepare_packet(data, WRITE_CHANNEL);
    return request;
}
